packages/sdk/src/protocols/base_protocol.ts

```diff
@@ -1,15 +1,206 @@
+import type { Peer, PeerId } from "@libp2p/interface";
+import { ConnectionManager } from "@waku/core";
+import { BaseProtocol } from "@waku/core/lib/base_protocol";
 import { IBaseProtocolSDK } from "@waku/interfaces";
+import { Logger } from "@waku/utils";
 
 interface Options {
   numPeersToUse?: number;
+  maintainPeersInterval?: number;
 }
 
 const DEFAULT_NUM_PEERS_TO_USE = 3;
+const DEFAULT_MAINTAIN_PEERS_INTERVAL = 30_000;
 
 export class BaseProtocolSDK implements IBaseProtocolSDK {
-  public readonly numPeers: number;
+  public readonly numPeersToUse: number;
+  private peers: Peer[] = [];
+  private maintainPeersIntervalId: NodeJS.Timeout | undefined;
+  log = new Logger("sdk:base-protocol");
 
-  constructor(options: Options) {
-    this.numPeers = options?.numPeersToUse ?? DEFAULT_NUM_PEERS_TO_USE;
+  maintainPeersLock = false;
+
+  constructor(
+    protected core: BaseProtocol,
+    private connectionManager: ConnectionManager,
+    options: Options
+  ) {
+    this.log = new Logger(`sdk:${core.multicodec}`);
+    this.numPeersToUse = options?.numPeersToUse ?? DEFAULT_NUM_PEERS_TO_USE;
+    const maintainPeersInterval =
+      options?.maintainPeersInterval ?? DEFAULT_MAINTAIN_PEERS_INTERVAL;
+
+    this.startMaintainPeersInterval(maintainPeersInterval).catch((error) => {
+      this.log.error("Error starting maintain peers interval:", error);
+      throw error;
+    });
+  }
+
+  get connectedPeers(): Peer[] {
+    return this.peers;
+  }
+
+  /**
+   * Disconnects from a peer and tries to find a new one to replace it.
+   * @param peerToDisconnect The peer to disconnect from.
+   */
+  public async renewPeer(peerToDisconnect: PeerId): Promise<void> {
+    this.log.info(`Renewing peer ${peerToDisconnect}`);
+    try {
+      await this.connectionManager.dropConnection(peerToDisconnect);
+      this.peers = this.peers.filter((peer) => peer.id !== peerToDisconnect);
+      this.log.info(
+        `Peer ${peerToDisconnect} disconnected and removed from the peer list`
+      );
+
+      await this.findAndAddPeers();
+    } catch (error) {
+      this.log.error(`Error renewing peer ${peerToDisconnect}:`, error);
+      this.log.info(
+        "Peer renewal failed, relying on the interval to find a new peer"
+      );
+    }
+  }
+
+  /**
+   * Stops the maintain peers interval.
+   */
+  public stopMaintainPeersInterval(): void {
+    if (this.maintainPeersIntervalId) {
+      this.log.info("Stopping maintain peers interval");
+      clearInterval(this.maintainPeersIntervalId);
+      this.maintainPeersIntervalId = undefined;
+      this.log.info("Maintain peers interval stopped");
```

Comment:
nit: it doesn't seem it can throw here so I would have only one log line and in general try to use less `log.info` as they are less informative and it can become too difficult to navigate logs later on